control.go

```diff
@@ -363,6 +363,37 @@ func (c *controlConn) reconnect() {
 		ch.conn.Close()
 	}
 
+	conn, err := c.attemptReconnectToAnyOfHosts(hosts)
+
+	if conn == nil {
+		c.session.logger.Printf("gocql: control falling back to initial contact points.\n")
+		// Fallback to initial contact points, as it may be the case that all known initial_hosts
+		// changed their IPs while keeping the same hostname(s).
+		initial_hosts, resolvErr := addrsToHosts(c.session.cfg.Hosts, c.session.pool.port, c.session.logger)
+		if resolvErr == nil {
+			conn, err = c.attemptReconnectToAnyOfHosts(initial_hosts)
+		} else {
+			c.session.logger.Printf("gocql: error on resolving contact points' hostnames: %v\n", err)
+		}
+	}
+
+	if conn == nil {
+		c.session.logger.Printf("gocql: control unable to register events: %v\n", err)
```

Comment:
It seems to me that currently we can log an earlier error (from `conn, err := c.attemptReconnectToAnyOfHosts(hosts)`) later than an error from the resolving (since err is not updated), which might be confusing.

It seems that inverting the error handling logic of the function might make the flow clearer:

```go
func (c *controlConn) reconnect() {
	if atomic.LoadInt32(&c.state) == controlConnClosing {
		return
	}
	if !atomic.CompareAndSwapInt32(&c.reconnecting, 0, 1) {
		return
	}
	defer atomic.StoreInt32(&c.reconnecting, 0)

	conn, err := c.attemptReconnect()

	if conn == nil {
		c.session.logger.Printf("gocql: unable to reconnect control connection: %v\n", err)
		return
	}

	err = c.session.refreshRing()
	if err != nil {
		c.session.logger.Printf("gocql: unable to refresh ring: %v\n", err)
	}
}

func (c *controlConn) attemptReconnect() (*Conn, error) {
	hosts := c.session.ring.allHosts()
	hosts = shuffleHosts(hosts)

	// keep the old behavior of connecting to the old host first by moving it to
	// the front of the slice
	ch := c.getConn()
	if ch != nil {
		for i := range hosts {
			if hosts[i].Equal(ch.host) {
				hosts[0], hosts[i] = hosts[i], hosts[0]
				break
			}
		}
		ch.conn.Close()
	}

	conn, err := c.attemptReconnectToAnyOfHosts(hosts)

	if conn != nil {
		return conn, err
	}

	c.session.logger.Printf("gocql: unable to connect to any ring node: %v\n", err)
	c.session.logger.Printf("gocql: control falling back to initial contact points.\n")
	// Fallback to initial contact points, as it may be the case that all known initialHosts
	// changed their IPs while keeping the same hostname(s).
	initialHosts, resolvErr := addrsToHosts(c.session.cfg.Hosts, c.session.cfg.Port, c.session.logger)
	if resolvErr != nil {
		return nil, fmt.Errorf("resolve contact points' hostnames: %v", resolvErr)
	}

	return c.attemptReconnectToAnyOfHosts(initialHosts)
}
```

What do you think?

Reply:
Definitely clearer. I'll refactor as you suggest.